lib/handlers/handler.js

```
@@ -28,19 +30,50 @@ MsgHandler.prototype.doHandle = function(sender, msg) {
+    // Message handled successfully, in case a result was given send it back.
+    if (!!res) {
+      sender.reply(msg, res);
+    }
     console.info('[MsgHandler] acks a msg with tag: ', msg.fields.consumerTag);
     sender.ack(msg);
-  }, function(err) {
-    if (sender.nackExchange) {
-      console.error('[MsgHandler] app-nacks a msg with tag: ', msg.fields.consumerTag);
-      sender.appNack(msg);
+  }).catch(function (err) {
+    console.error('[MsgHandler] recived the following error from handler: ' + err);
+    // If error is NOT of the temporary type, reply as error, otherwise reinsert it back to the message queue
+    if (!(err instanceof TemporaryError)) {
+      sender.error(msg, (err && err.message) ? err.message :
+      'Handler did not provide appropriate error, err obj was: ' + err);
+      console.info('[MsgHandler] acks a msg with tag: ', msg.fields.consumerTag);
+      sender.ack(msg);
     } else {
-      // Uses Rabbit's default nack, can't handle retries
-      console.error('[MsgHandler] nacks a msg with tag: ', msg.fields.consumerTag);
-      sender.nack(msg);
+      if (sender.nackExchange) {
+        console.error('[MsgHandler] app-nacks a msg with tag: ', msg.fields.consumerTag);
+        sender.appNack(msg);
+      } else {
+        console.error('[MsgHandler] nacks a msg with tag: ', msg.fields.consumerTag);
+        sender.nack(msg);
+      }
     }
   });
 };
 
+MsgHandler.prototype.handleJson = function (msg) {
+  var msgBody = msg.content.toString();
+  console.info(`[MsgHandler] handles msg:\n${msgBody}`);
+
+  console.log('msgBody', msgBody);
+  var parsedMessage;
+  try {
+    parsedMessage = JSON.parse(msgBody);
+  } catch(e) {
+    console.error(`[${this.constructor.name}] handler expected a message in a JSON format, \
+but received the following: \n${msgBody}
+original error message was: ${e.message}`);
+    e.message = `${this.constructor.name} received a bad formatted msg`;
```

I think it's a bad idea to change error objects, would be better to throw a new error instead